# Algorithm integration

This workflow integrates all variant calls in a batch of samples on a
per-algorithm basis.

Frequently, samples are called jointly in a small group, which is a subset of a
larger batch of sequencing libraries. To accomodate this paradigm, the user
specifies a sample key, indicating which group and batch each sample belongs
to, e.g.

```
sample	group	batch
11194.fa	11194	pcr_plus
11006.mo	11006	pcr_minus
```

The workflow expects a PE/SR VCF for each group, and a depth bed for each
batch, as detailed below, and integrates all variants belonging to a single
batch. A group may be a single sample; simply use the sample ID as the group
ID.

## Input files

Variant calls must be standardized to the following specifications before
integration. See the example preprocessing module for guidance.

* `pesr/input_vcfs/{source}.{group}.vcf.gz`  
    Tabix-indexed PE/SR VCFs per algorithm. 
    - {source} indicates the source algorithm.
    - {group} indicates a subgroup of samples which were called jointly.
    - VCF records are required to include INFO fields for SVTYPE, CHR2, END,
      STRANDS [++,+-,-+,--], SVLEN, and SOURCES.  Currently, DEL, DUP, INV, and
      BND are supported SVTYPEs; INS have not been tested. BND and INV
      breakpoints must be segregated and annotated with strand.  
      (NOTE: END has become a reserved attribute in pysam 0.11.2.1 and may be
      deprecated here in the future.)
* `depth/input_beds/{batch}.{svtype}.bed.gz`  
    All per-sample depth calls in a batch, merged across algorithms.
    - Depth calls must be segregated by {svtype}, which may be DEL or DUP.
    - Each line corresponds to a single call in a single sample.
    - First six columns must be chrom, start, end, name, sample, svtype. Any
      following columns are discarded during integration.
    - In our preprocessing module, these files are generated by taking all
      algorithms' calls for a given sample and performing a bedtools merge,
      then concatenating and sorting all sample calls.

## Output files

* `pesr/vcfcluster/{batch}.{source}.{chrom}.vcf`  
    Clustered PE/SR variants, per batch, per source, and per chromosome.
* `pesr/rdtest_beds/{batch}.{source}.{chrom}.bed`  
    Clustered PE/SR variants in RdTest format. 
* `depth/rdtest_beds/{batch}.depth.{chrom}.bed`  
    Clustered depth variants, per batch and per chromosome, in RdTest format.

## Configuration

All variables controlling pipeline operation can be modified in `config.yaml`.
This documentation is incomplete and in progress.

* `batches` : filepath  
    Sample/group/batch key.

* `groups` : filepath  
    List of groups to include during integration. Expects one PE/SR VCF per
    group.

* `chroms`: filepath  
    List of chromosomes to include during integration. Integration is
    parallelized by chromosome.

* `pesr_sources` : list of strings  
    List of all PE/SR algorithms to merge. Expects one VCF from each algorithm
    for each group; integrating an algorithm from only a subset of groups is
    not yet supported.

* `cnv_types`: list of strings  
    List of SV classes to include during integration of depth variants.
    Mostly useful for testing. (Default: DEL,DUP)

* `vcfcluster` : list of params  
    vcfcluster parameters; see `svtools vcfcluster -h` for details

* `bedcluster` : list of params  
    bedcluster parameters; see `svtools bedcluster -h` for details
